cores/arduino/stm32/bootloader.c

```diff
@@ -2,7 +2,144 @@
 
+/*
+ * STM32 built-in bootloader in system memory support
+ */
+/* Private definitions to manage system memory address */
+#define SYSMEM_ADDR_COMMON 0xFFF
+
+typedef struct {
+  uint32_t devID;
+  uint32_t sysMemAddr;
+} devSysMemAddr_str;
+
+devSysMemAddr_str devSysMemAddr[] = {
+#ifdef STM32F0xx
+  {0x440, 0x1FFFEC00},
+  {0x444, 0x1FFFEC00},
+  {0x442, 0x1FFFD800},
+  {0x445, 0x1FFFC400},
+  {0x448, 0x1FFFC800},
+  {0x442, 0x1FFFD800},
+#elif STM32F1xx
+  {0x412, 0x1FFFF000},
+  {0x410, 0x1FFFF000},
+  {0x414, 0x1FFFF000},
+  {0x420, 0x1FFFF000},
+  {0x428, 0x1FFFF000},
+  {0x418, 0x1FFFB000},
+  {0x430, 0x1FFFE000},
+#elif STM32F2xx
+  {0x411, 0x1FFF0000},
+#elif STM32F3xx
+  {SYSMEM_ADDR_COMMON, 0x1FFFD800},
+#elif STM32F4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32F7xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32G0xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32G4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32H7xx
+  {0x450, 0x1FF00000},
+#elif STM32L0xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32L1xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32L4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32WBxx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#else
+#warning "No system memory address for this serie!"
+#endif
+  {0x0000, 0x00000000}
+};
+
+uint32_t getSysMemAddr(void)
+{
+  uint32_t sysMemAddr = 0;
+  if (devSysMemAddr[0].devID == SYSMEM_ADDR_COMMON) {
+    sysMemAddr = devSysMemAddr[0].sysMemAddr;
+  } else {
+    uint32_t devId = LL_DBGMCU_GetDeviceID();
+    for (uint32_t id = 0; devSysMemAddr[id].devID != 0; id++) {
+      if (devSysMemAddr[id].devID == devId) {
+        sysMemAddr = devSysMemAddr[id].sysMemAddr;
+        break;
+      }
+    }
+  }
+  return sysMemAddr;
+}
+
+/* Request to jump to system memory boot */
+WEAK void jumpToBootloaderRequested(void)
+{
+  enableBackupDomain();
+  setBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX, SYSBL_MAGIC_NUMBER_BKP_VALUE);
+  NVIC_SystemReset();
+}
+
+/* Jump to system memory boot from user application */
+WEAK void jumpToBootloader(void)
+{
+  enableBackupDomain();
+  if (getBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX) == SYSBL_MAGIC_NUMBER_BKP_VALUE) {
+    setBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX, 0);
+
+#ifdef USBCON
+    USBD_reenumerate();
+#endif
+    void (*sysMemBootJump)(void);
+
+    /**
+     * Get system memory address
+     *
+     * Available in AN2606 document:
+     * Table xxx Bootloader device-dependent parameters
+     */
+    volatile uint32_t sysMem_addr = getSysMemAddr();
+    if (sysMem_addr != 0) {
+#ifdef __HAL_SYSCFG_REMAPMEMORY_SYSTEMFLASH
+      /* Remap system Flash memory at address 0x00000000 */
+      __HAL_SYSCFG_REMAPMEMORY_SYSTEMFLASH();
+#endif
+
+      /**
+       * Set jump memory location for system memory
+       * Use address with 4 bytes offset which specifies jump location
+       * where program starts
+       */
+      sysMemBootJump = (void (*)(void))(*((uint32_t *)(sysMem_addr + 4)));
```

Comment:
The first word is the stack pointer value. The next on is the resetVector.

Reply:
That's not quite what I meant. A few lines up, `sysMem_addr` is remapped onto 0x0 (as the MCU also does when starting with the BOOT0 pin), so the reset vector is available at `sysMem_addr + 4`, but now also at `4`. I wonder if this should jump to the remapped address `4` instead (since I think that is also what the hardware does when BOOT0 is used).

Reply:
Ah OK. That's can make sense. Anyway I could not answer. Will try to find the answer.

Reply:
I thought about this a bit more, and I think it does not actually matter: The jump is not *to* 0x04 (or `sysMem_addr + 4`), it only loads the jump address from there. I guess it does not really matter which of these two addresses is used for this load (the resulting value to jump to will be the same).

One advantage of loading from 0x4 after remapping, is that you do not actually need to know where the system flash lives, which can simplify the code. I have already implemented this, see comments further down in this PR.